src/dmd/escape.d

```diff
@@ -33,6 +37,199 @@ import dmd.tokens;
+     * passed by ref to the function (which they essentially are via the static link).
+     */
+    VarDeclaration[] outerVars = fd ? fd.outerVars[] : null;
+
+    const len = arguments.length + (ethis !is null) + outerVars.length;
+    if (len <= 1)
+        return errors;
+
+    struct EscapeBy
+    {
+        EscapeByResults er;
+        Parameter param;        // null if no Parameter for this argument
+        bool isMutable;         // true if reference to mutable
+    }
+
+    /* Store escapeBy as static data escapeByStorage so we can keep reusing the same
+     * arrays rather than reallocating them.
+     */
+    __gshared EscapeBy[] escapeByStorage;
+    auto escapeBy = escapeByStorage;
+    if (escapeBy.length < len)
+    {
+        auto newPtr = cast(EscapeBy*)mem.xrealloc(escapeBy.ptr, len * EscapeBy.sizeof);
+        // Clear the new section
+        memset(newPtr + escapeBy.length, 0, (len - escapeBy.length) * EscapeBy.sizeof);
+        escapeBy = newPtr[0 .. len];
+        escapeByStorage = escapeBy;
+    }
+
+    const paramLength = tf.parameterList.length;
+
+    // Fill in escapeBy[] with arguments[], ethis, and outerVars[]
+    foreach (const i, ref eb; escapeBy)
+    {
+        bool refs;
+        Expression arg;
+        if (i < arguments.length)
+        {
+            arg = (*arguments)[i];
+            if (i < paramLength)
+            {
+                eb.param = tf.parameterList[i];
+                refs = (eb.param.storageClass & (STC.out_ | STC.ref_)) != 0;
+                eb.isMutable = eb.param.isReferenceToMutable(arg.type);
+            }
+            else
+            {
+                eb.param = null;
+                refs = false;
+                eb.isMutable = arg.type.isReferenceToMutable();
+            }
+        }
+        else if (ethis)
+        {
+            /* ethis is passed by value if a class reference,
+             * by ref if a struct value
+             */
+            eb.param = null;
+            arg = ethis;
+            auto ad = fd.isThis();
+            assert(ad);
+            assert(ethis);
+            if (ad.isClassDeclaration())
+            {
+                refs = false;
+                eb.isMutable = arg.type.isReferenceToMutable();
+            }
+            else
+            {
+                assert(ad.isStructDeclaration());
+                refs = true;
+                eb.isMutable = arg.type.isMutable();
+            }
+        }
+        else
+        {
+            // outer variables are passed by ref
+            eb.param = null;
+            refs = true;
+            auto var = outerVars[i - (len - outerVars.length)];
+            eb.isMutable = var.type.isMutable();
+            eb.er.byref.push(var);
+            continue;
+        }
+
+        if (refs)
+            escapeByRef(arg, &eb.er);
+        else
+            escapeByValue(arg, &eb.er);
+    }
+
+    foreach (const i, ref eb; escapeBy[0 .. $ - 1])
```

Comment:
So this is... `O((n^2)/2 * (v+r))` in the normal case (no error) (`n == escapeBy.length`, `v == byValue.length`, `r == byRef.length`). That doesn't sound quite right.

Reply:
It's bad if there are a lot of parameters, but those cases are awfully rare.

Reply:
Since there is no usage of this yet, we don't know how impactful it will be. But what we already know is that it doesn't scale at all.